File: sdk-typescript/src/services/nfs.ts
/**
 * NFS Service for mounting sandbox workspaces
 */

import { execSync, spawn, ChildProcess } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

/**
 * Options for NFS mount
 */
export interface NfsMountOptions {
  /** NFS server host */
  host: string;
  /** NFS server port (default: 2049) */
  port?: number;
  /** Export path on the server */
  exportPath: string;
  /** Local mount point (auto-created if not specified) */
  mountPoint?: string;
  /** Custom mount options (default: nfsvers=3,tcp,nolock,port={port},mountport={port}) */
  options?: string;
}

/**
 * Represents an active NFS mount
 */
export class NfsMount {
  private readonly host: string;
  private readonly port: number;
  private readonly exportPath: string;
  private _mountPoint: string;
  private readonly options: string;
  private _mounted: boolean = false;
  private _tempDir: string | null = null;

  constructor(options: NfsMountOptions) {
    this.host = options.host;
    this.port = options.port ?? 2049;
    this.exportPath = options.exportPath;
    this._mountPoint = options.mountPoint ?? '';
    this.options = options.options ?? 'nfsvers=3,tcp,nolock,port={port},mountport={port}';
  }

  /**
   * Get the mount point path
   */
  get mountPoint(): string {
    if (this._mountPoint) {
      return this._mountPoint;
    }
    throw new Error('Mount point not initialized');
  }

  /**
   * Check if currently mounted
   */
  get isMounted(): boolean {
    return this._mounted;
  }

  /**
   * Mount the NFS share
   * @returns The mount point path
   */
  mount(): string {
    if (this._mounted) {
      return this._mountPoint;
    }

    // Create mount point if not specified
    if (!this._mountPoint) {
      this._tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'workspace_nfs_'));
      this._mountPoint = this._tempDir;
    } else {
      fs.mkdirSync(this._mountPoint, { recursive: true });
    }

    // Build mount options
    const opts = this.options.replace(/{port}/g, String(this.port));

    // Mount command
    const cmd = `mount -t nfs -o ${opts} ${this.host}:${this.exportPath} ${this._mountPoint}`;

    try {
      execSync(cmd, { stdio: 'pipe' });
      this._mounted = true;
      return this._mountPoint;
    } catch (error: any) {
      if (this._tempDir) {
        fs.rmdirSync(this._tempDir);
        this._tempDir = null;
      }
      throw new Error(`Failed to mount NFS: ${error.stderr?.toString() || error.message}`);
    }
  }

  /**
   * Unmount the NFS share
   */
  unmount(): void {
    if (!this._mounted) {
      return;
    }

    try {
      execSync(`umount ${this._mountPoint}`, { stdio: 'pipe' });
    } catch {
      // Try lazy unmount
      try {
        execSync(`umount -l ${this._mountPoint}`, { stdio: 'pipe' });
      } catch {
        // Ignore errors
      }
    } finally {
      this._mounted = false;
      if (this._tempDir) {
        try {
          fs.rmdirSync(this._tempDir);
        } catch {
          // Ignore cleanup errors
        }
        this._tempDir = null;
      }
    }
  }
}

/**
 * Service for managing NFS mounts for sandbox workspaces
 */
export class NfsService {
  private readonly defaultHost?: string;
  private readonly defaultPort: number;

  /**
   * Create an NFS service
   * @param defaultHost Default NFS server host
   * @param defaultPort Default NFS port (default: 2049)
   */
  constructor(defaultHost?: string, defaultPort: number = 2049) {
    this.defaultHost = defaultHost;
    this.defaultPort = defaultPort;
  }

  /**
   * Create an NFS mount for a sandbox
   * @param sandboxId The sandbox ID to mount
   * @param options Mount options
   * @returns NfsMount instance
   *
   * @example
   * ```typescript
   * const mount = nfs.mount('sandbox-123');
   * try {
   *   mount.mount();
   *   // Access files at mount.mountPoint
   *   fs.writeFileSync(`${mount.mountPoint}/test.txt`, 'Hello');
   * } finally {
   *   mount.unmount();
   * }
   * ```
   */
  mount(
    sandboxId: string,
    options?: {
      mountPoint?: string;
      host?: string;
      port?: number;
    }
  ): NfsMount {
    const host = options?.host ?? this.defaultHost;
    const port = options?.port ?? this.defaultPort;

    if (!host) {
      throw new Error('NFS host not specified and no default configured');
    }

    return new NfsMount({
      host,
      port,
      exportPath: `/${sandboxId}`,
      mountPoint: options?.mountPoint,
    });
  }

  /**
   * Create an NFS mount from a full NFS URL
   * @param nfsUrl NFS URL in format nfs://host:port/path
   * @param mountPoint Optional local mount point
   * @returns NfsMount instance
   */
  mountFromUrl(nfsUrl: string, mountPoint?: string): NfsMount {
    const url = new URL(nfsUrl);
    const host = url.hostname;
    const port = url.port ? parseInt(url.port, 10) : this.defaultPort;
    const exportPath = url.pathname;

    return new NfsMount({
      host,
      port,
      exportPath,
      mountPoint,
    });
  }

  /**
   * Check if NFS mount is available on this system
   */
  static isAvailable(): boolean {
    try {
      execSync('which mount.nfs', { stdio: 'pipe' });
      return true;
    } catch {
      return false;
    }
  }
}

File: sdk-python/src/workspace_sdk/services/nfs.py
"""
NFS service for mounting sandbox workspaces
"""

import os
import subprocess
import tempfile
from pathlib import Path
from typing import Optional
from urllib.parse import urlparse


class NfsMount:
    """Context manager for NFS mount"""

    def __init__(
        self,
        host: str,
        port: int,
        export_path: str,
        mount_point: Optional[str] = None,
        options: Optional[str] = None,
    ):
        self.host = host
        self.port = port
        self.export_path = export_path
        self._mount_point = mount_point
        self._temp_dir: Optional[tempfile.TemporaryDirectory] = None
        self._mounted = False
        self.options = options or "nfsvers=3,tcp,nolock,port={port},mountport={port}"

    @property
    def mount_point(self) -> str:
        """Get the mount point path"""
        if self._mount_point:
            return self._mount_point
        if self._temp_dir:
            return self._temp_dir.name
        raise RuntimeError("Mount point not initialized")

    def mount(self) -> str:
        """
        Mount the NFS share.

        Returns:
            The mount point path
        """
        if self._mounted:
            return self.mount_point

        # Create mount point if not specified
        if not self._mount_point:
            self._temp_dir = tempfile.TemporaryDirectory(prefix="workspace_nfs_")
            mount_path = self._temp_dir.name
        else:
            mount_path = self._mount_point
            Path(mount_path).mkdir(parents=True, exist_ok=True)

        # Build mount options
        opts = self.options.format(port=self.port)

        # Mount command
        cmd = [
            "mount",
            "-t", "nfs",
            "-o", opts,
            f"{self.host}:{self.export_path}",
            mount_path,
        ]

        try:
            subprocess.run(cmd, check=True, capture_output=True, text=True)
            self._mounted = True
            return mount_path
        except subprocess.CalledProcessError as e:
            if self._temp_dir:
                self._temp_dir.cleanup()
                self._temp_dir = None
            raise RuntimeError(f"Failed to mount NFS: {e.stderr}") from e

    def unmount(self) -> None:
        """Unmount the NFS share"""
        if not self._mounted:
            return

        try:
            subprocess.run(
                ["umount", self.mount_point],
                check=True,
                capture_output=True,
                text=True,
            )
        except subprocess.CalledProcessError as e:
            # Try lazy unmount
            try:
                subprocess.run(
                    ["umount", "-l", self.mount_point],
                    check=True,
                    capture_output=True,
                    text=True,
                )
            except subprocess.CalledProcessError:
                pass
        finally:
            self._mounted = False
            if self._temp_dir:
                self._temp_dir.cleanup()
                self._temp_dir = None

    def __enter__(self) -> "NfsMount":
        self.mount()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.unmount()


class NfsService:
    """Service for managing NFS mounts for sandbox workspaces"""

    def __init__(self, default_host: Optional[str] = None, default_port: int = 2049):
        """
        Initialize NFS service.

        Args:
            default_host: Default NFS server host (can be overridden per mount)
            default_port: Default NFS port (default: 2049)
        """
        self.default_host = default_host
        self.default_port = default_port

    def mount(
        self,
        sandbox_id: str,
        mount_point: Optional[str] = None,
        host: Optional[str] = None,
        port: Optional[int] = None,
        nfs_url: Optional[str] = None,
    ) -> NfsMount:
        """
        Create an NFS mount for a sandbox.

        Args:
            sandbox_id: The sandbox ID to mount
            mount_point: Optional mount point path (auto-created if not specified)
            host: NFS server host (uses default if not specified)
            port: NFS server port (uses default if not specified)
            nfs_url: Full NFS URL (overrides host/port/sandbox_id if provided)

        Returns:
            NfsMount context manager

        Example:
            with nfs.mount("sandbox-123") as mount:
                # Access files at mount.mount_point
                with open(f"{mount.mount_point}/test.txt", "w") as f:
                    f.write("Hello")
        """
        if nfs_url:
            # Parse nfs://host:port/path URL
            parsed = urlparse(nfs_url)
            host = parsed.hostname or self.default_host
            port = parsed.port or self.default_port
            export_path = parsed.path
        else:
            host = host or self.default_host
            port = port or self.default_port
            export_path = f"/{sandbox_id}"

        if not host:
            raise ValueError("NFS host not specified and no default configured")

        return NfsMount(
            host=host,
            port=port,
            export_path=export_path,
            mount_point=mount_point,
        )

    @staticmethod
    def is_available() -> bool:
        """Check if NFS mount is available on this system"""
        try:
            result = subprocess.run(
                ["which", "mount.nfs"],
                capture_output=True,
                text=True,
            )
            return result.returncode == 0
        except Exception:
            return False
